design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,150 @@
+```
+
+### End to End Workflow
+
+#### Changes in Common.yaml
+
+Following contents need to be append to `common.yaml` :
+
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>
+
+#### ceph-object-cosi-controller
+
+The `ceph-object-cosi-controller` will span Ceph COSI Driver pod with default settings on the Rook Operator Namespace. The controller will bring up the Ceph COSI driver when first object store is created and will stop the COSI driver when last object store is deleted only if it detect COSI Controller is running on default namespace. The controller will also watch for CephCosiDriver CRD, if defined the driver will be started with the settings provided in the CRD. If the Ceph COSI driver if up and running,it  will also create `CephObjectStoreUser` named  `cosi` for each object store which internally creates a secret rook-ceph-object-user-<objectstore-name>-cosi provides credentials for the object store. This can be specified in the BucketClass and BucketAccessClass. Also this controller ensures maximum one CephCosiDriver CRD exists in the cluster. For v1.12 supports object store belong to RGW multisite are not supported.
+
+#### cephcosidriver.ceph.rook.io CRD
+
+The users can define following CRD so that configuration related the Ceph COSI driver can be passed. This is not mandatory and the driver will be started with default settings if this CRD is not defined.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCosiDriver
+metadata:
+  name: rook-ceph-cosi
+  namespace: rook-ceph
+spec:
+  deploymentStrategy: "Auto"
+```
+
+The `deploymentStrategy` can be `Auto` or `Force` or `Disable`. If `Auto` is specified the driver will be started automatically when first object store is created and will be stopped when last object store is deleted. If `Force` is specified the driver will be started when the CRD is created and will be stopped when the CRD is deleted. The user can also disable the driver by specifying `Disable` as the value.
+
+#### Creating COSI Related CRDs
+
+There are five different kubernetes resources related to COSI. They are `Bucket`, `BucketAccess`, `BucketAccessClass`, `BucketClass` and `BucketClaim`. The user can create these resources using following `kubectl` command. All the examples will be added to deploy/examples/cosi directory in the Rook repository.
+
+```bash
+kubectl create -f deploy/examples/cosi/bucketclass.yaml
+kubectl create -f deploy/examples/cosi/bucketclaim.yaml
+kubectl create -f deploy/examples/cosi/bucketaccessclass.yaml
+kubectl create -f deploy/examples/cosi/bucketaccess.yaml
+```
+
+User need refer Secret which contains access credentials for the object store in the `Parameter`  field of `BucketClass` and `BucketAccessClass` CRD as below:
+
+```yaml
+Parameter
+  credentialsReference: "rook-ceph-object-user-<objectstore-name>-cosi"
+  namespaceReference: "<objectstore-store-namespace>"
```

Comment:
Do we have control of these parameter names? If so, what about `objectStoreSecretName` and `objectStoreNamespace`?

Reply:
Yes we have, but I was trying for a more generic hence didn't append `objectStore`

Reply:
We could put more than just the credentials in the secret, right? So it seems helpful to have the descriptive name so they know what secret and namespace should be used. Perhaps `objectUserSecretName` and `objectStoreNamespace`